DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,235 @@
+            _count = _count > 0 ? --_count : _count;
+            return result;
+        }
+
+        #region IEnumerable Implementation
+        public IEnumerator<T> GetEnumerator() 
+        {
+            for (int i = _start; i < Count; i++) 
+            {
+                yield return _circularBuffer[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() 
+        {
+            return GetEnumerator();
+        }
+        #endregion
+
+        #region ICollection Implementation
+        private int _count;
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get 
+            {
+                return _count;
+            }
+        }
+        /// <summary>
+        /// Checks whether this collection is readonly
+        /// </summary>
+        public bool IsReadOnly 
+        {
+            get 
+            {
+                return false;
+            }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            _count = 0;
+            _start = 0;
+            _end = 0;
+            _circularBuffer = new T[Length + 1];
+        }
+        /// <summary>
+        /// Checks whether the buffer contains an item
+        /// </summary>
+        public bool Contains(T item) 
+        {
+            return _circularBuffer.Contains(item);
+        }
+        /// <summary>
+        /// Copies this buffer to an array
+        /// </summary>
+        public void CopyTo(T[] array, int arrayIndex) 
+        {
+            if (array == null) 
+            {
+                throw new ArgumentNullException("array can not be null");
+            }
+
+            if (array.Length == 0 || arrayIndex >= array.Length || arrayIndex < 0) 
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            // Get enumerator
+            var enumarator = GetEnumerator();
+
+            // Copy elements if there is any in the buffer and if the index is within the valid range
+            while (arrayIndex < array.Length) 
+            {
+                if (enumarator.MoveNext()) 
+                {
+                    array[arrayIndex] = enumarator.Current;
+                    arrayIndex++;
+                }
+                else 
+                {
+                    break;
+                }
+            }
+        }
+        /// <summary>
+        /// Removes an item from the buffer
+        /// </summary>
+        public bool Remove(T item) 
+        {
+            if (!IsEmpty && Contains(item)) 
+            {
+               var sourceArray = _circularBuffer.Except(new T[] { item }).ToArray();
+                _circularBuffer = new T[Length + 1];
+                Array.Copy(sourceArray, _circularBuffer, sourceArray.Length);
+
+                if (!item.Equals(default(T))) 
```

Comment:
I think this `if` statement is not needed, but I might be wrong, so can you please add a test for removing duplicate `default(T)` values from a buffer? Here's an example case:

```
{ 1, 0, 2, 0, 3, 0, 4, 0, 5, *0*, *0*, *0* }
```

Where, `*0*` is an empty slot, but `0` is not, and was actually added by the user.

Reply:
Please refer to the following REPL, I think there are several problems with the new implementation of `Remove()`, please extend the test cases to also cover removing duplicate entries that are `default(T)` for both integers and strings.

REPL: https://repl.it/@aalhour/csharp-circular-buffer-pr-review

Reply:
Apologies for not replying earlier.

You are absolutely right, we need to update the `_end `and `_count `all the time. But, we need the `if` statement. Actually we need an `if-else` statement ( at least, I think it's easier to understand ).

Reply:
Here's how I modified it currently.
                ```
if (!Equals(item,default(T))) 
                {
                    _end = sourceArray.Length - 1;
                    _count = sourceArray.Length-1;
                }
                else 
                {
                    _end = sourceArray.Length;
                    _count = sourceArray.Length;
                }
```
Lets consider a buffer of length 12:  `{1,0,2,0,3,0,4,0,5,*0*,*0*,*0*,*0*}`. Internally, it's of length 13 as explained earlier.

When we remove `0`, `sourceArray` becomes `{1,2,3,4,5}` and `_circularBuffer` becomes {1,2,3,4,5,*0*,*0*,*0*,*0*,*0*,*0*,*0*,*0*}. This clearly shows that `sourceArray.Length` is the correct `_count` of `_circularBuffer`. The next value should be inserted at `index=5`. Thus `_end` is still equal to the `sourceArray.Length`.

However, when we are removing item that is not equal to `default(T)`, these calculations have to change because of the `<empty space>` that will be left in the `sourceArray`.

That's why I used the `if-else` construct. This implementation currently passes all the tests including that of strings. Please review.